core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1799 @@
+    override val byte: Byte = 0x02.byteValue
+  }
+
+  final case object SighashTypeKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x03.byteValue
+  }
+
+  final case object RedeemScriptKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x04.byteValue
+  }
+
+  final case object WitnessScriptKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x05.byteValue
+  }
+
+  final case object BIP32DerivationPathKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x06.byteValue
+  }
+
+  final case object FinalizedScriptSigKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x07.byteValue
+  }
+
+  final case object FinalizedScriptWitnessKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x08.byteValue
+  }
+
+  final case object ProofOfReservesCommitmentKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x09.byteValue
+  }
+
+  final case object UnknownKeyId extends PSBTInputKeyId {
+    override val byte: Byte = Byte.MaxValue
+  }
+}
+
+sealed trait PSBTOutputKeyId extends PSBTKeyId
+
+object PSBTOutputKeyId {
+
+  def fromByte(byte: Byte): PSBTOutputKeyId = byte match {
+    case RedeemScriptKeyId.byte        => RedeemScriptKeyId
+    case WitnessScriptKeyId.byte       => WitnessScriptKeyId
+    case BIP32DerivationPathKeyId.byte => BIP32DerivationPathKeyId
+    case _: Byte                       => UnknownKeyId
+  }
+
+  def fromBytes(key: ByteVector): PSBTOutputKeyId = {
+    if (key.isEmpty) {
+      UnknownKeyId
+    } else {
+      fromByte(key.head)
+    }
+  }
+
+  final case object RedeemScriptKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x00.byteValue
+  }
+
+  final case object WitnessScriptKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x01.byteValue
+  }
+
+  final case object BIP32DerivationPathKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x02.byteValue
+  }
+
+  final case object UnknownKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = Byte.MaxValue
+  }
+}
+
+sealed trait PSBTMap extends NetworkElement {
+  require(elements.map(_.key).groupBy(identity).values.forall(_.length == 1),
+          "All keys must be unique.")
+
+  def elements: Vector[PSBTRecord]
+
+  def bytes: ByteVector =
+    elements
+      .sortBy(_.key)
+      .foldLeft(ByteVector.empty)(_ ++ _.bytes) ++ hex"00"
+}
+
+case class GlobalPSBTMap(elements: Vector[GlobalPSBTRecord]) extends PSBTMap {
+
+  def getRecords[T <: GlobalPSBTRecord](key: PSBTGlobalKeyId): Vector[T] = {
+    elements
+      .filter(element => PSBTGlobalKeyId.fromByte(element.key.head) == key)
+      .asInstanceOf[Vector[T]]
+  }
+
+  /**
+    * Takes another GlobalPSBTMap and adds all records that are not contained in this GlobalPSBTMap
+    * @param other GlobalPSBTMap to be combined with
+    * @return A GlobalPSBTMap with the combined data of the two GlobalPSBTMaps
+    */
+  def combine(other: GlobalPSBTMap): GlobalPSBTMap = {
```

Comment:
should this function start with a require that the `UnsignedTransaction`s within are the same?

Reply:
wouldn't want to end up with two of them